### ISSUE TITLE
event system?

### ISSUE THREAD

**Issue by smallka** [NSSTC/sim-ecs#34]
refer to bevy: https://github.com/bevyengine/bevy/blob/latest/examples/ecs/event.rs

**Comment by minecrawler** [comment]
I wanted to get events into sim-ecs for quiet a while, and I have been thinking about ways to make them a reality. sim-ecs is not built the same way bevy_ecs works (for a number of reasons), so there are some design concerns I have and I can't find a good solution. So, let me just brainstorm:

**Option 1**
ReadEvent and WriteEvent are part of the regular IAccessQuery interface, but Read only triggers the query when there is an event / the query only has data when there is an event

- does every data-instance of varying components get the same reference to the event data struct?
- how do we handle multiple events?
- sounds inconsistent with regular way queries work
- might need looping system execution to handle all events

```
class MyEvent {
    constructor(
        public message: string,
    ) {}
}

const query = new Query({
    eventReader: ReadEvent(MyEvent),
    eventWriter: WriteEvent(MyEvent),
});

for (const {eventReader, eventWriter} of query.iter()) {
    console.log(eventReader.message); // <- ref is constant for all items in query
    setTimeout(eventWriter.send(new MyEvent('Hello World')), 100); // <- ref is constant for all items in query
}
```


**Option 2**
Expose a new interface IEventSystem, which is a special System interface for handling events without touching the current Query APIs

```
class MyEvent {
    constructor(
        public message: string,
    ) {}
}
class MyEventSystem extends EventSystem {
    readonly query = new Query({ counterObj: Write(Counter) });

    run() {
        this.query.execute(({counterObj}) => {
          console.log(++counterObj.a);
        });
    }
}

```



**Option 3**
Implement an API on the world interfaces which allows manually polling events at any point

**Comment by minecrawler** [comment]
After creating a PoC with option 4, it seems to be a good move forward, so I will implement centralized pipelines, now, and after that work on stageless, System builder and finally events. Might take a while, but I'll try to work on it quickly and release it as a package in 0.5.0

**Comment by minecrawler** [comment]
Merged in #43 and will be part of v0.5
